### PR TITLE
[ARCTIC-1423][FLINK]: OffsetOutOfRangeException is encountered when the arctic logstore failover.

### LINKED ISSUE #1423: [Bug]: OffsetOutOfRangeException is encountered when the arctic logstore failover.
### What happened?

When Flink Job failover, Arctic Logstore will encounter OffsetOutOfRangeException which shows 'Offsets out of range with no configured reset policy for partitions'.

### Affects Versions

master

### What engines are you seeing the problem on?

Flink

### How to reproduce

1. Start a Flink Job that consumers from Arctic Logstore
2. Let this Job failover in some way, for example by restarting ams
3. OffsetOutOfRangeException will occur.

### Relevant log output

```shell
2023-05-04 16:57:00
java.lang.RuntimeException: One or more fetchers have encountered exception
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcherManager.checkErrors(SplitFetcherManager.java:199)
	at org.apache.flink.connector.base.source.reader.SourceReaderBase.getNextFetch(SourceReaderBase.java:154)
	at org.apache.flink.connector.base.source.reader.SourceReaderBase.pollNext(SourceReaderBase.java:116)
	at org.apache.flink.streaming.api.operators.SourceOperator.emitNext(SourceOperator.java:275)
	at org.apache.flink.streaming.runtime.io.StreamTaskSourceInput.emitNext(StreamTaskSourceInput.java:67)
	at org.apache.flink.streaming.runtime.io.StreamOneInputProcessor.processInput(StreamOneInputProcessor.java:65)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.processInput(StreamTask.java:398)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:191)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:619)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:583)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:758)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:573)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: SplitFetcher thread 0 received unexpected exception while polling the records
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.runOnce(SplitFetcher.java:146)
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.run(SplitFetcher.java:101)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more
Caused by: org.apache.kafka.clients.consumer.OffsetOutOfRangeException: Offsets out of range with no configured reset policy for partitions: {}
	at org.apache.kafka.clients.consumer.internals.Fetcher.parseCompletedFetch(Fetcher.java:1221)
	at org.apache.kafka.clients.consumer.internals.Fetcher.fetchedRecords(Fetcher.java:591)
	at org.apache.kafka.clients.consumer.KafkaConsumer.pollForFetches(KafkaConsumer.java:1294)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1225)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1201)
	at com.netease.arctic.flink.read.source.log.kafka.LogKafkaPartitionSplitReader.readNormal(LogKafkaPartitionSplitReader.java:135)
	at com.netease.arctic.flink.read.source.log.kafka.LogKafkaPartitionSplitReader.fetch(LogKafkaPartitionSplitReader.java:126)
	at org.apache.flink.connector.base.source.reader.fetcher.FetchTask.run(FetchTask.java:56)
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.runOnce(SplitFetcher.java:138)
	... 6 more
```

### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #1431: [Bug]: when read from Logstore, will encounter ClassCastException.
### What happened?

when read from logstore, flink job will fail and shows ClassCastException.

### Affects Versions

master

### What engines are you seeing the problem on?

Flink

### How to reproduce

_No response_

### Relevant log output

```shell
2023-05-09 20:30:45
java.lang.ClassCastException: com.netease.arctic.flink.read.source.log.kafka.LogKafkaPartitionSplitReader cannot be cast to org.apache.flink.connector.kafka.source.reader.KafkaPartitionSplitReader
	at com.netease.arctic.flink.read.internals.KafkaSourceFetcherManager.enqueueOffsetsCommitTask(KafkaSourceFetcherManager.java:90)
	at com.netease.arctic.flink.read.internals.KafkaSourceFetcherManager.commitOffsets(KafkaSourceFetcherManager.java:77)
	at com.netease.arctic.flink.read.internals.KafkaSourceReader.notifyCheckpointComplete(KafkaSourceReader.java:133)
	at org.apache.flink.streaming.api.operators.SourceOperator.notifyCheckpointComplete(SourceOperator.java:307)
	at org.apache.flink.streaming.runtime.tasks.StreamOperatorWrapper.notifyCheckpointComplete(StreamOperatorWrapper.java:99)
	at org.apache.flink.streaming.runtime.tasks.SubtaskCheckpointCoordinatorImpl.notifyCheckpointComplete(SubtaskCheckpointCoordinatorImpl.java:319)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.notifyCheckpointComplete(StreamTask.java:1089)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.lambda$notifyCheckpointCompleteAsync$11(StreamTask.java:1054)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.lambda$notifyCheckpointOperation$13(StreamTask.java:1077)
	at org.apache.flink.streaming.runtime.tasks.StreamTaskActionExecutor$1.runThrowing(StreamTaskActionExecutor.java:50)
	at org.apache.flink.streaming.runtime.tasks.mailbox.Mail.run(Mail.java:90)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.processMail(MailboxProcessor.java:317)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:189)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:619)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:583)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:758)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:573)
	at java.lang.Thread.run(Thread.java:748)
```

### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by lklhdu**
## Why are the changes needed?
fix #1423 which results in OffsetOutOfRangeException.
fix #1431. When performing the unit test in 1.14 and 1.15 modules, ClassCastException related to #1431 was encountered and should be fixed.

## Brief change log

  - *Modified the parameters passed to the LogKafkaPartitionSplit constructor*
  - *Override the KafkaSourceFetcherManager class from flink-kafka-connector to fix ClassCastException*
  - *The flink 1.12, 1.14 and 1.15 versions are affected.*

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink/v1.12/flink/src/test/java/com/netease/arctic/flink/read/hidden/kafka/TestKafkaSourceReader.java [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+    } while (output.count() < TOTAL_NUM_RECORDS);
+
+    // The completion of the last checkpoint should subsume all the previous checkpoints.
+    assertEquals(checkpointId, reader.getOffsetsToCommit().size());
+    reader.notifyCheckpointComplete(checkpointId);
+
+    // re-create and restore
+    reader = (LogKafkaSourceReader) createReader(groupId);
+    reader.addSplits(splitList);
+    List<KafkaPartitionSplit> currentSplitList = reader.snapshotState(checkpointId++);
```
Comment (YesOrNo828) [inline flink/v1.12/flink/src/test/java/com/netease/arctic/flink/read/hidden/kafka/TestKafkaSourceReader.java:123]: Why `checkpoint++` here?  I think only `checkpoint`, WDYT?

File: flink/v1.12/flink/src/test/java/com/netease/arctic/flink/read/hidden/kafka/TestKafkaSourceReader.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+  }
+
+  @AfterClass
+  public static void shutdown() throws Exception {
+    KAFKA_CONTAINER.close();
+  }
+
+  @Before
+  public void initData() throws Exception {
+    topic = TestUtil.getUtMethodName(testName);
```
Comment (YesOrNo828) [inline flink/v1.12/flink/src/test/java/com/netease/arctic/flink/read/hidden/kafka/TestKafkaSourceReader.java:97]: How many partitions does this topic have?

**Review by YesOrNo828 — COMMENTED** [review]
Does this issue #1431 affect on flink-1.14 and 1.15 versions?

File: flink/v1.12/flink/src/test/java/com/netease/arctic/flink/read/hidden/kafka/TestKafkaSourceReader.java (hunk 1 above)
Comment (lklhdu) [reply flink/v1.12/flink/src/test/java/com/netease/arctic/flink/read/hidden/kafka/TestKafkaSourceReader.java]: Yes, you're right, I've modified it.

File: flink/v1.12/flink/src/test/java/com/netease/arctic/flink/read/hidden/kafka/TestKafkaSourceReader.java (hunk 2 above)
Comment (lklhdu) [reply flink/v1.12/flink/src/test/java/com/netease/arctic/flink/read/hidden/kafka/TestKafkaSourceReader.java:98]: The default number of partitions is 1, which I have changed to be explicitly specified in the unit test.

**Comment by lklhdu** [comment]
> Does this issue #1431 affect on flink-1.14 and 1.15 versions?

This issue affects all three modules, but the 1.12-related one has been fixed before.

**Review by YesOrNo828 — APPROVED** [review]
LGTM

**Review by YesOrNo828 — APPROVED** [review]
LGTM

**Pull request merged**